Fixes for wxSpinCtrl::GetSizeFromTextSize()

vadz opened the pull request:
This works better (i.e. without leaving excess space for GTK 3) for me with both GTK 2 and 3 and seems reasonable, but I'd still be grateful if you could please look at these changes, Paul. TIA!

vadz committed:
Return fitting size from wxSpinCtrl::GetSizeFromTextSize() in GTK

Previously we added the extent of the text to the preferred size of a
spin button with 0 digits, but this didn't account for the fact that the
control still reserved enough space for 3 digits in this case, and so
the returned size was always too big than the required size by the width
needed to show 3 digits.

Fix this by directly asking for the size required to show the number of
digits corresponding to the given text size.

Note that ideal would be to provide GTK API at wx level directly, i.e.
allow to just specify the number of digits instead of the text extent.
But we still need to fix the already existing GetSizeFromTextSize()
anyhow, so do this for now.

See #18568.

vadz committed:
Don't change max number of digits in wxSpinCtrl code

This doesn't seem to affect anything, i.e. the function returns the same
results whether we do it or not, so just don't.

wangqr commented on src/gtk/spinctrl.cpp:
```diff
@@ -364,27 +365,31 @@ wxSize wxSpinCtrlGTKBase::DoGetSizeFromTextSize(int xlen, int ylen) const
+    // string of digits in the first place, but there doesn't seem to be
+    // anything better to do (unless we add some GetSizeFromNumberOfDigits()).
+    const double widthDigit = GetTextExtent("0123456789").GetWidth() / 10.0;
+    const int numDigits = wxRound(xlen / widthDigit);
```
Can we override GetSizeFromText and just strlen the input text?

vadz replied:
It's currently not virtual, but it could be indeed made virtual and overridden here. Is this really better than adding `GetSizeFromNumberOfDigits()`? Not sure... in any case, this is rather low priority IMO, because, as I wrote in the ticket, neither should be necessary anyhow as `wxSpinCtrl` is supposed to determine its best size correctly automatically.

paulcor commented on src/gtk/spinctrl.cpp:
```diff
@@ -364,26 +365,32 @@ wxSize wxSpinCtrlGTKBase::DoGetSizeFromTextSize(int xlen, int ylen) const
     gtk_entry_set_width_chars(GTK_ENTRY(m_widget), widthChars);
 
-    wxSize tsize(xlen + totalS.x, totalS.y);
+    // Somehow, gtk_widget_get_preferred_size() doesn't account for the borders
```
If this were true, there would be a problem with the height as well. What is happening here is `widthChars == -1`, so the actual preferred size is being computed using the upper and lower bounds of the range (see the GTK3 private function `gtk_spin_button_get_text_width()`). It seems that this is sometimes turning out to be just a couple pixels bigger than the preferred size we're getting for `numDigits`.

vadz replied:
Why is `widthChars == -1` though when we explicitly set it to 0 in the current master and `numDigits` in this PR?

Also, it does happen with the height too (preferred height is 36, allocated height is 34), but I just ignored it because the height is much bigger than the font height anyhow, so it doesn't matter. I probably shouldn't be lazy and do it for the height as well.

This being said, I'm still not sure about what's really going on here and if this has anything to do with `wxSpinButton::DoGetBestSize()`, where I don't understand why is it removing padding from the best width neither...

paulcor replied:
> Why is widthChars == -1 though when we explicitly set it to 0 in the current master and numDigits in this PR?

I meant literally the `widthChars` variable. Try this:
```diff
--- src/gtk/spinctrl.cpp-	2019-11-12 07:49:10.089190178 -0800
+++ src/gtk/spinctrl.cpp	2019-11-12 07:49:39.890547001 -0800
@@ -375,8 +375,10 @@
     gtk_entry_set_width_chars(GTK_ENTRY(m_widget), numDigits);
 
     wxSize tsize = GTKGetPreferredSize(m_widget);
+fprintf(stderr, "%d", tsize.x);
 
     gtk_entry_set_width_chars(GTK_ENTRY(m_widget), widthChars);
+fprintf(stderr, " %d\n", GTKGetPreferredSize(m_widget).x);
 
     // Somehow, gtk_widget_get_preferred_size() doesn't account for the borders
     // with this widget, and if we use it as is, we get GTK errors when
```
> Also, it does happen with the height too (preferred height is 36, allocated height is 34)

I don't see any evidence of this. The reported preferred height is 34, and the control is perfectly happy being sized to that, but if you size it to 33, it complains.
>wxSpinButton::DoGetBestSize(), where I don't understand why is it removing padding from the best width neither...

Because the padding is for the text part of the control, which we don't want to show for `wxSpinButton`, see also 135e3875e2.

vadz replied:
OK, I see, thanks. I've actually tried calling `GTKGetPreferredSize()` with different char width values under debugger, so I did see this, but didn't think it through.

However I am still not sure what should be done about this: should we just call `gtk_entry_set_width_chars()` in ctor/`DoSetRange()`? I.e. instead of leaving it as default -1 always set it to the actually needed value?

BTW, am I correct in thinking that there is no difference between `PROP_WIDTH_CHARS` and `PROP_MAX_WIDTH_CHARS` in our case? AFAICS the former is used to compute the min size and the latter -- the "natural" size, but they're the same for `wxSpinCtrl`.

paulcor replied:
> should we just call gtk_entry_set_width_chars() in ctor/DoSetRange()?

Probably.
> am I correct in thinking that there is no difference between PROP_WIDTH_CHARS and PROP_MAX_WIDTH_CHARS in our case?

Also probably. As long as nothing sets `max-width-chars`.

vadz committed:
Move private wxSpinCtrl helpers into a private header

Np real changes, just don't put these functions in the public
wx/spinctrl.h, they have nothing to do there.

vadz committed:
Rename wxSpinCtrl helper functions

Put them in wxSpinCtrlImpl namespace and remove "wxSpinCtrl" prefix from
the function names themselves, this was ugly.

No real changes.

vadz committed:
Factor out another wxSpinCtrl helper for getting its max length

This will be used in wxGTK implementation.

For now no real changes yet.

vadz committed:
Adjust entry width of wxSpinCtrl in wxGTK to its range

Ensure that the entry is always (just) big enough to show any value
valid in this spin control.

This also ensures that GetBestSize() doesn't need to be overridden to
use GetSizeFromTextSize() any longer as the best size will be determined
correctly by GTK itself.

vadz commented:
After adding calls to `gtk_entry_set_width_chars()` the hack with border is indeed not needed any longer, thanks for noticing and explaining this. Moreover, now we don't even need to override `DoGetBestSize()`, which is nice.

Please let me know if you see any problems with the newest version. TIA!
